mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala

```diff
@@ -315,12 +315,17 @@ class ALS(override val uid: String) extends Estimator[ALSModel] with ALSParams {
 
   override def fit(dataset: DataFrame): ALSModel = {
     import dataset.sqlContext.implicits._
-    val ratings = dataset
-      .select(col($(userCol)).cast(IntegerType), col($(itemCol)).cast(IntegerType),
-        col($(ratingCol)).cast(FloatType))
-      .map { row =>
-        Rating(row.getInt(0), row.getInt(1), row.getFloat(2))
+    val ratings = if ($(ratingCol) != "") {
```

Comment:
It might be easier to define the rating col first:

~~~scale
val r = if ($(ratingCol) != "") col($(ratingCol)).cast(FloatType) else lit(1.0f)
val ratings = dataset.select(..., ..., r).map { row =>
  ...
}
~~~